### PR TITLE
Feature/search

### PR DESCRIPTION
**Pull request by katherinebernstein**

### CODE REVIEW

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+			initializeEntitySetDataModelIndex();
+		}
+	}
+	
+	@Override
+	public Boolean initializeEntitySetDataModelIndex() {
+		try {
+			if ( !verifyElasticsearchConnection() ) return false;
+		} catch (UnknownHostException e) {
+			e.printStackTrace();
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: Log an error don't just leave print strack trace.

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+		try {
+			String s = mapper.writeValueAsString( entitySetDataModel );
+			client.prepareIndex( ENTITY_SET_DATA_MODEL, ENTITY_SET_TYPE, entitySet.getId().toString() ).setSource( s ).execute().actionGet();
+			updateEntitySetPermissions(
+					entitySet.getId(),
+					principal,
+					Sets.newHashSet( Permission.OWNER, Permission.READ, Permission.WRITE, Permission.DISCOVER, Permission.LINK ) );
+			return true;
+		} catch (JsonProcessingException e) {
+			e.printStackTrace();
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: Another print stack trace

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+	@Override
+	public List<Map<String, Object>> executeEntitySetDataModelKeywordSearch(
+			Optional<String> optionalSearchTerm,
+			Optional<UUID> optionalEntityType,
+			Optional<Set<UUID>> optionalPropertyTypes,
+			Set<Principal> principals ) {
+		try {
+			if ( !verifyElasticsearchConnection() ) return Lists.newArrayList();
+		} catch (UnknownHostException e) {
+			e.printStackTrace();
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: Ditto

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+		ObjectMapper mapper = new ObjectMapper();
+		mapper.registerModule( new GuavaModule() );
+		mapper.registerModule( new JodaModule() );
+		try {
+			String s = mapper.writeValueAsString( acl );
+			String id = entitySetId.toString() + "_" + principal.getType().toString() + "_" + principal.getId();
+			client.prepareIndex( ENTITY_SET_DATA_MODEL, ACLS, id ).setParent( entitySetId.toString() ).setSource( s ).execute().actionGet();
+			return true;
+		} catch (JsonProcessingException e) {
+			e.printStackTrace();
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: ditto

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+		}
+		return false;
+	}
+	
+	@Override
+	public Boolean updatePropertyTypesInEntitySet( UUID entitySetId, Set<PropertyType> newPropertyTypes ) {
+		try {
+			if ( !verifyElasticsearchConnection() ) return false;
+		} catch (UnknownHostException e) {
+			e.printStackTrace();
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: ditto

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+		ObjectMapper mapper = new ObjectMapper();
+		mapper.registerModule( new GuavaModule() );
+		mapper.registerModule( new JodaModule() );
+		try {
+			String s = mapper.writeValueAsString( propertyTypes );
+			UpdateRequest updateRequest = new UpdateRequest( ENTITY_SET_DATA_MODEL, ENTITY_SET_TYPE, entitySetId.toString() ).doc( s );
+			client.update( updateRequest ).get();
+			return true;
+		} catch (IOException | InterruptedException | ExecutionException e) {
+			e.printStackTrace();
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: ditto

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+	public Boolean updatePropertyTypesInEntitySet( UUID entitySetId, Set<PropertyType> newPropertyTypes ) {
+		try {
+			if ( !verifyElasticsearchConnection() ) return false;
+		} catch (UnknownHostException e) {
+			e.printStackTrace();
+		}
+		
+		Map<String, Object> propertyTypes = Maps.newHashMap();
+		propertyTypes.put( PROPERTY_TYPES, newPropertyTypes);
+		ObjectMapper mapper = new ObjectMapper();
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: Use ObjectMappers if you can to get your object mapper. They are expensive to setup and are designed to be used by multiple threads safely.

File: src/main/java/com/kryptnostic/kindling/search/ElasticsearchTransportClientFactory.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+import org.elasticsearch.common.transport.InetSocketTransportAddress;
+import org.elasticsearch.transport.client.PreBuiltTransportClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+
+public class ElasticsearchTransportClientFactory {
+	
+	public static final Logger log = LoggerFactory.getLogger( ElasticsearchTransportClientFactory.class );
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ElasticsearchTransportClientFactory.java]: We use logger pretty much everywhere, including earlier in this PR. Switch to logger here for consistency.

File: src/main/java/com/kryptnostic/kindling/search/ElasticsearchTransportClientFactory.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		);
+		
+		if ( verifyConnection( client ) ) {
+			return client;
+		} else {
+			return null;
+		}
+	}
+	
+	public boolean verifyConnection( Client someClient ) {
```
Comment (geekbeast) [inline src/main/java/com/kryptnostic/kindling/search/ElasticsearchTransportClientFactory.java]: minor: I would call this `isConnected(...)`

File: src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java (hunk 7 above)
Comment (katherinebernstein) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: I updated this to only initialize mapper once--is this what you meant?

Comment (katherinebernstein) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: jk figured it out

Comment (hochung) [inline src/main/java/com/kryptnostic/kindling/search/ConductorElasticsearchImpl.java]: there's a factory class `ObjectMappers` in rhizome-client

**Pull request merged**